examples/chat/module/rust/src/lib.rs

```diff
@@ -14,137 +14,76 @@
-    fn create_room(&mut self, req: CreateRoomRequest) -> grpc::Result<Empty> {
-        info!("creating room");
-        if self.rooms.contains_key(&req.room_id) {
-            return room_id_duplicate_err();
-        }
-
-        // Create a new Node for this room, and keep the write handle and admin token in the `rooms`
-        // map.
-        self.rooms.insert(req.room_id, Room::new(req.admin_token));
-
-        Ok(Empty::new())
-    }
-
-    fn destroy_room(&mut self, req: DestroyRoomRequest) -> grpc::Result<Empty> {
-        info!("destroying room");
-        match self.rooms.entry(req.room_id) {
-            Entry::Occupied(e) => {
-                if e.get().admin_token == req.admin_token {
-                    // Close the only input channel that reaches the per-room Node, which
-                    // will trigger it to terminate.
-                    e.get().sender.close().expect("could not close channel");
-                    e.remove();
-                    Ok(Empty::new())
-                } else {
-                    Err(grpc::build_status(
-                        grpc::Code::PERMISSION_DENIED,
-                        "invalid admin token",
-                    ))
-                }
-            }
-            Entry::Vacant(_) => room_id_not_found_err(),
+impl Router {
+    fn event_loop(mut self, in_handle: u64) -> Result<(), oak::OakError> {
+        let in_channel = oak::ReadHandle {
+            handle: oak::Handle::from_raw(in_handle),
+        };
+        let receiver = oak::io::Receiver::new(in_channel);
+        info!("starting event loop");
+        loop {
+            let grpc_invocation: Invocation = receiver.receive()?;
+            self.handle_grpc_invocation(grpc_invocation)?;
         }
     }
 
-    fn subscribe(&mut self, req: SubscribeRequest, mut writer: grpc::ChannelResponseWriter) {
-        info!("subscribing to room");
-        match self.rooms.get(&req.room_id) {
-            None => {
-                writer
-                    .close(room_id_not_found_err())
-                    .expect("could not close channel");
-            }
-            Some(room) => {
-                info!("new subscription to room {:?}", req.room_id);
-                let command = Command::Join(writer.handle());
-                room.sender
-                    .send(&command)
-                    .expect("could not send command to room Node");
-            }
-        };
+    fn handle_grpc_invocation(&mut self, grpc_invocation: Invocation) -> Result<(), oak::OakError> {
+        let grpc_invocation_label = grpc_invocation.request_receiver.label()?;
+        info!(
+            "handling incoming gRPC invocation with label {:?}",
+            grpc_invocation_label
+        );
+        let room_node_sender = self
+            .rooms
+            .entry(grpc_invocation_label.clone())
+            .or_insert_with(|| create_room_node(&grpc_invocation_label));
+        room_node_sender.send(&grpc_invocation)
```

Comment:
This feels like it provides a very specific solution that moves the problem elsewhere (forwarding the gRPC channel to another room) rather than providing a generic mechanism that can be used for other routing use cases where the incoming channel into the node is not gRPC.

I assume the more generic solution will come later with a more complete label system implementation.

Reply:
Fair point, I think this would be addressed (or at least facilitated) by making it easier for client to specify labels for each call, as opposed to only once per channel. WDYT?

Reply:
Yes, per-call labels should help to resolve this.